### PR TITLE
module registry observer

### PR DESCRIPTION
**Pull request by mikea**
enables recording timing/memory/usage information

EW PR 5129

### CODE REVIEW

File: src/workerd/jsg/modules.h [hunk 1]
```diff
@@ -79,6 +79,24 @@ v8::Local<v8::WasmModuleObject> compileWasmModule(jsg::Lock& js, auto&& reader)
 
 void instantiateModule(jsg::Lock& js, v8::Local<v8::Module>& module);
 
+enum class ModuleInfoCompileOption {
+  BUNDLE,
+  // The BUNDLE options tells the compile operation to threat the content as coming
```
Comment (ohodson) [inline src/workerd/jsg/modules.h:84]: s/threat/treat/

Comment (ohodson) [reply src/workerd/jsg/modules.h:84]: (existing spello)

File: src/workerd/jsg/modules.c++ [hunk 2]
```diff
@@ -306,7 +306,10 @@ v8::Local<v8::Module> compileEsmModule(
     jsg::Lock& js,
     kj::StringPtr name,
     kj::ArrayPtr<const char> content,
-    ModuleInfoCompileOption option) {
+    ModuleInfoCompileOption option,
+    const ModuleRegistryObserver& observer) {
+  auto compilationObserver = observer.onEsmCompilationStart(js.v8Isolate, name, option);
```
Comment (jasnell) [inline src/workerd/jsg/modules.c++:311]: Adding a comment about how `compilationObserver` is expected to be "used" here would be helpful to future generations.

**Review by jasnell — APPROVED** [review]

File: src/workerd/jsg/modules.c++ [hunk 3]
```diff
@@ -306,7 +306,11 @@ v8::Local<v8::Module> compileEsmModule(
     jsg::Lock& js,
     kj::StringPtr name,
     kj::ArrayPtr<const char> content,
-    ModuleInfoCompileOption option) {
+    ModuleInfoCompileOption option,
+    const ModuleRegistryObserver& observer) {
+  // destroy the observer after compilation finished to indicate the end of the process.
+  auto compilationObserver = observer.onEsmCompilationStart(js.v8Isolate, name, option);
```
Comment (ohodson) [inline src/workerd/jsg/modules.c++:312]: If I understand what could be happening here: an observer returns an owned<void> which is an opaque container (`kj::Own<void>`) for a class / struct that "observes" the start and end of compilation events when the opaque container is created and destroyed. A `ModuleRegisryObserver` implementation can choose what happens on start and end of compilation. The LHS is doing the observing the end of the compilation on behalf of the `ModuleRegistryObserver`, e.g. compilationEndObserver.

I might call `compilationObserver` something like `compilationEvent` / `compilationLifetime`

Comment (mikea) [reply src/workerd/jsg/modules.c++:312]: Thanks. I thought about it a little bit, and decided to keep the name observer.

Couple reasons:
- we already use destructor in observers to send an event, so this usage/naming is consistent
- this observer is very simple now (only destructor), but may get additional methods in the future or got rid of the destructor altogether if we need to send more information (e.g. successful/non-successful)

**Pull request merged**